torchmultimodal/modules/layers/transformer_decoder.py

```diff
@@ -0,0 +1,199 @@
+    ):
+        super().__init__()
+        if (
+            position_embedding_input.embedding_dim
+            != position_embedding_output.embedding_dim
+        ):
+            raise ValueError(
+                f"The position embedding dimension of the input modality {position_embedding_input.embedding_dim}"
+                f"must be equal to that of the output modality {position_embedding_output.embedding_dim}"
+            )
+
+        self.embedding_dim = position_embedding_input.embedding_dim
+        self.right_shift = RightShift(self.embedding_dim)
+        self.position_embedding_input = position_embedding_input
+        self.position_embedding_output = position_embedding_output
+        self.input_seq_len = self.position_embedding_input.seq_len
+        self.output_seq_len = self.position_embedding_output.seq_len
+        total_len = self.input_seq_len + self.output_seq_len
+
+        self.attention_nets = nn.ModuleList(
+            [
+                TransformerLayer(
+                    shape=(total_len),
+                    embedding_dim=self.embedding_dim,
+                    num_attention_heads=num_attention_heads,
+                    num_hidden_layers=num_hidden_layers,
+                    dropout=dropout,
+                    attention_dropout=attention_dropout,
+                )
+                for i in range(num_hidden_layers)
+            ]
+        )
+
+    def _get_inference_embeddings(self, x, position_embedding, decode_step):
+        # Use the data point at ``decode_step - 1`` to generate the next.
+        return position_embedding(x, decode_step)
+
+    def _get_training_embeddings(self, x, position_embedding):
+        # Use the full sequence for training
+        return x + position_embedding(x)
+
+    def forward(self, x_input, x_output, decode_step=None, decode_idx=None):
+        # Trigger inference mode if either of the input/output-modality sequence is ``None`` as generation
+        # is based the previous data point (autoregressive) where only one modality is needed at any point
+        # along the sequence.
+        if x_output is None:
+            # Skip inferencing if ``decode_step = 0`` as decoding requires previous data point.
+            # ``decode_step`` ranges between ``(0, input_seq_len]`` controlled by the GPT model.
+            # When ``decode_step = input_seq_len``, the first output modality is generated from the last input
+            # modality.
+            if decode_step > 0:
+                x_input = self._get_inference_embeddings(
+                    x_input, self.position_embedding_input, decode_step
+                )
+            x = x_input
+        elif x_input is None:
+            # Continue to generate output-modality sequence from ``input_seq_len + 1``(inclusive) to
+            # ``input_seq_len + output_seq_len - 1``(inclusive).
+            x = self._get_inference_embeddings(
+                x_output,
+                self.position_embedding_output,
```

Comment:
if we want this to become a generic decoder, we need to push this preprocessing to the caller.